sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -249,6 +251,58 @@ def test_absolute_error_sample_weight():
     gbdt.fit(X, y, sample_weight=sample_weight)
 
 
+@pytest.mark.parametrize("y", [([1.0, -2.0, 0.0]), ([0.0, 1.0, 2.0])])
+def test_gamma_y_positive(y):
+    # Test that ValueError is raised if any y_i <= 0.
+    err_msg = r"loss='gamma' requires positive y."
+    gbdt = HistGradientBoostingRegressor(loss="gamma", random_state=0)
+    with pytest.raises(ValueError, match=err_msg):
+        gbdt.fit(np.zeros(shape=(len(y), 1)), y)
+
+
+def test_gamma():
+    # For Gamma distributed target, Gamma loss should give better results
+    # than least squares or Poisson measured in Gamma deviance as metric.
+    rng = np.random.RandomState(42)
+    n_train, n_test, n_features = 500, 500, 20
+    X = make_low_rank_matrix(
+        n_samples=n_train + n_test,
+        n_features=n_features,
+        random_state=rng,
+    )
+    # We create a log-linear Gamma model. This gives y.min ~ 1e-2, y.max ~ 1e2
+    coef = rng.uniform(low=-10, high=20, size=n_features)
+    # Numpy parametrizes gamma(shape=k, scale=theta) with mean = k * theta and
+    # variance = k * theta^2. We want parametrized instead with mean = exp(X @ coef)
+    # and variance = dispersion * mean^2 by setting k = 1 / dispersion,
+    # theta =  dispersion * mean.
+    dispersion = 0.5
+    y = rng.gamma(shape=1 / dispersion, scale=dispersion * np.exp(X @ coef))
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=n_test, random_state=rng
+    )
+    gbdt_gamma = HistGradientBoostingRegressor(loss="gamma", random_state=123)
+    gbdt_ls = HistGradientBoostingRegressor(loss="squared_error", random_state=123)
+    gbdt_pois = HistGradientBoostingRegressor(loss="poisson", random_state=123)
+    for model in (gbdt_gamma, gbdt_ls, gbdt_pois):
+        model.fit(X_train, y_train)
+    dummy = DummyRegressor(strategy="mean").fit(X_train, y_train)
+
+    # Improve unconditional calibration on the training set by a correction factor.
+    # This almost always improves out-of-sample predictive accuracy.
+    cor = np.mean(y_train) / np.mean(gbdt_gamma.predict(X_train))
```

Comment:
Shall we had an hyper-parameter to `HistGradientBoostingRegressor` do this automatically? (In a dedicated PR?)

Reply:
Or is this only expected to work for the gamma loss?

Reply:
I just read:

> Open question: In practice, I (almost) always observed that a post-train calibration of gradient boosted trees with a gamma loss is beneficial not only for calibration but also for out-of-sample performance. The step is to add a multiplicative constant (or in link-space additive constant = reset _baseline_prediction) such that on the training set one has y_train.mean() == model.predict(X_train).mean(). Should we include it by default or add an option or not consider it (which would make me sad)?

I would be in favor of an option.

Reply:
I'd like such an option as well. I'll open an issue or PR for it.

Reply:
I've not discussed this with other libraries. But it is of practical importance and I do it anyway. So having it as feature would be really nice. Maybe, I'll open similar issues in those libraries a little later:wink:

Reply:
I opened #22435.